File: selenium-java/src/test/java/com/maelstromops/qa/pages/CustomersPage.java
package com.maelstromops.qa.pages;

import com.maelstromops.qa.support.Config;
import org.openqa.selenium.By;
import org.openqa.selenium.WebDriver;

public class CustomersPage extends BasePage {

  private final By root = By.cssSelector("[data-testid='customer.browse']");
  private final By firstViewButton =
      By.xpath("(//button[normalize-space()='View'])[1]");

  public CustomersPage(WebDriver driver) {
    super(driver);
  }

  public CustomersPage open() {
    driver.get(Config.baseUrl() + "/customer/browse");
    visible(root);
    return this;
  }

  public boolean isLoaded() {
    return isVisible(root);
  }

  /** Click the first row's "View" command — navigates to that customer's profile. */
  public void openFirstProfile() {
    clickable(firstViewButton).click();
  }
}

File: playwright-ts/src/pages/CustomersPage.ts
import { expect, type Locator, type Page } from '@playwright/test';

/** Page Object for the Customers browse grid. */
export class CustomersPage {
  readonly page: Page;
  readonly root: Locator;

  constructor(page: Page) {
    this.page = page;
    this.root = page.getByTestId('customer.browse');
  }

  async goto(): Promise<void> {
    await this.page.goto('/customer/browse');
  }

  async expectLoaded(): Promise<void> {
    await expect(this.root).toBeVisible();
  }

  /** Click the first row's "View" command button — opens that customer's profile. */
  async openFirstProfile(): Promise<void> {
    await this.page.getByRole('button', { name: 'View' }).first().click();
  }
}
